miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/workflow/DefaultWorkflowManager.java

```diff
@@ -0,0 +1,146 @@
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.Workflow;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.WorkflowExecutor;
+import uk.ac.bbsrc.tgac.miso.core.data.workflow.impl.ProgressImpl;
+import uk.ac.bbsrc.tgac.miso.core.store.ProgressStore;
+import uk.ac.bbsrc.tgac.miso.service.exception.ValidationError;
+import uk.ac.bbsrc.tgac.miso.service.exception.ValidationException;
+import uk.ac.bbsrc.tgac.miso.service.security.AuthorizationManager;
+import uk.ac.bbsrc.tgac.miso.service.workflow.factory.BarcodableProgressStepFactory;
+import uk.ac.bbsrc.tgac.miso.service.workflow.factory.IntegerProgressStepFactory;
+import uk.ac.bbsrc.tgac.miso.service.workflow.factory.ProgressStepFactory;
+
+@Service
+@Transactional(rollbackFor = Exception.class)
+public class DefaultWorkflowManager implements WorkflowManager {
+  @Autowired
+  private AuthorizationManager authorizationManager;
+
+  @Autowired
+  private ProgressStore progressStore;
+
+  @Autowired
+  private WorkflowExecutor workflowExecutor;
+
+  @Autowired
+  private BarcodableProgressStepFactory barcodableProgressStepFactory;
+
+  @Autowired
+  private IntegerProgressStepFactory integerProgressStepFactory;
+
+  @Override
+  public Workflow beginWorkflow(String workflowNameString) throws IOException {
+    WorkflowName workflowName = WorkflowName.valueOf(workflowNameString);
+
+    Progress progress = new ProgressImpl();
+    progress.setWorkflowName(workflowName);
+
+    saveProgress(progress);
+    return workflowName.createWorklow(progress);
+  }
+
+  @Override
+  public Workflow processInput(Workflow workflow, String input) throws IOException {
+    workflow.processInput(makeProgressStep(input, workflow.getNextStep().getDataTypes()));
+    saveProgress(workflow.getProgress());
+    return workflow;
+  }
+
+  @Override
+  public Workflow processInput(Workflow workflow, int stepNumber, String input) throws IOException {
+    workflow.processInput(stepNumber, makeProgressStep(input, workflow.getNextStep().getDataTypes()));
+    saveProgress(workflow.getProgress());
+    return workflow;
+  }
+
+  @VisibleForTesting
+  protected ProgressStep makeProgressStep(String input, Set<InputType> inputTypes) throws IOException {
+    for (FactoryType factoryType : getFactoryTypes(inputTypes)) {
+      ProgressStep step = makeFactory(factoryType).create(input, inputTypes);
+      if (step != null) {
+        return step;
+      }
+    }
+
+    throw new ValidationException(Collections.singletonList(new ValidationError("Could not construct ProgressStep")));
+  }
+
+  /**
+   * @return sorted, de-duplicated list of FactoryTypes
+   */
+  private List<FactoryType> getFactoryTypes(Set<InputType> inputTypes) {
+    return inputTypes.stream().map(InputType::getFactoryType).distinct().sorted().collect(Collectors.toList());
+  }
+
+  private ProgressStepFactory makeFactory(FactoryType factoryType) {
+    switch (factoryType) {
+    case BARCODABLE:
+      return barcodableProgressStepFactory;
```

Comment:
This method could be simplified by Autowiring a `Map<FactoryType, ProgressStepFactory` (see https://github.com/oicr-gsi/miso-lims/blob/9ef4254f7f56812af9ef91c62fbfed8bca8feb3e/miso-web/src/main/java/uk/ac/bbsrc/tgac/miso/webapp/context/WebConfig.java#L60 for the definition of a similar Map and https://github.com/oicr-gsi/miso-lims/blob/8afde626497c426366c6d9b66b4d68d9eb7a2732/miso-service/src/main/java/uk/ac/bbsrc/tgac/miso/service/impl/DefaultBarcodableViewService.java#L28 for an example of its usage).

I gave this a try, but this Autowiring will not work in the service layer tests, since it depends on `miso-web`, I believe.  Also, this kind of Map introduces complexity, and it would seem out of place in `WebConfig`.

Reply:
You need a Spring context, and the class to be autowired has to be a bean within that context for autowiring to happen. Usually you'd use mocks instead of autowiring things into the class being tested though.